docs/src/main/asciidoc/config-mappings.adoc

```diff
@@ -0,0 +1,456 @@
+- Support for `Map`, `List`, `Set`, `Optional` and primitive types.
+
+Additionally, the annotations `@WithName`, `@WithParentName`, `@WithDefault`, `@WithConverter` may be used to override
+the default behaviours of ConfigxMapping interfaces.
+
+=== Examples
+
+==== Simple Config Mapping
+[source,java]
+----
+interface Server {
+    String host();
+
+    int port();
+}
+----
+
+==== Nested Config Mapping
+
+[source,java]
+----
+interface ServerSub {
+    @WithParentName
+    ServerSubHostAndPort subHostAndPort();
+
+    @WithParentName
+    ServerSubName subName();
+}
+
+interface ServerSubHostAndPort {
+    String host();
+
+    int port();
+}
+
+interface ServerSubName {
+    String name();
+}
+----
+
+==== Types Config Mapping
+
+[source,java]
+----
+public interface SomeTypes {
+    @WithName("int")
+    int intPrimitive();
+
+    @WithName("int")
+    Integer intWrapper();
+
+    @WithName("long")
+    long longPrimitive();
+
+    @WithName("long")
+    Long longWrapper();
+
+    @WithName("float")
+    float floatPrimitive();
+
+    @WithName("float")
+    Float floatWrapper();
+
+    @WithName("double")
+    double doublePrimitive();
+
+    @WithName("double")
+    Double doubleWrapper();
+
+    @WithName("char")
+    char charPrimitive();
+
+    @WithName("char")
+    Character charWrapper();
+
+    @WithName("boolean")
+    boolean booleanPrimitive();
+
+    @WithName("boolean")
+    Boolean booleanWrapper();
+}
+----
+
+==== Optional Config Mapping
+
+[source,java]
+----
+public interface Optionals {
+    Optional<Server> server();
+
+    Optional<String> optional();
+
+    @WithName("optional.int")
+    OptionalInt optionalInt();
+}
+----
+
+==== Collections Config Mapping
+
+[source,java]
+----
+public interface CollectionTypes {
+    @WithName("strings")
+    List<String> listStrings();
+
+    @WithName("ints")
+    List<Integer> listInts();
+}
+
+public interface Maps {
+    @WithParentName
+    Map<String, Server> server();
+
+    Map<String, String> anotherServer();
+}
+
+public interface ServerCollections {
+    List<Environment> environments();
+
+    interface Environment {
+        String name();
+
+        List<App> apps();
+
+        interface App {
+            String name();
+
+            List<String> services();
+
+            Optional<List<String>> databases();
+        }
+    }
+}
+----
+
+==== Defaults Config Mapping
+
+[source,java]
+----
+public interface Defaults {
+    @WithDefault("foo")
+    String foo();
+
+    @WithDefault("bar")
+    String bar();
+}
+----
+
+==== Converter Config Mapping
+
+[source,java]
+----
+public interface Converters {
+    @WithConverter(FooBarConverter.class)
+    String foo();
+}
+
+public static class FooBarConverter implements Converter<String> {
+    @Override
+    public String convert(final String value) {
+        return "bar";
+    }
+}
+----
+
+[[config-properties]]
+== [.line-through]#@ConfigProperties# (Deprecated)
```

Comment:
If it's deprecated, I would just remove it from the documentation and add a line in the migration guide.

Reply:
Well, I was thinking that these should go through a deprecation cycle. This feature is somehow popular and a lot of people use it, so I'm unsure about removing them completely from the documentation, considering that we don't even have versions docs at the moment. My plan, was to just remove it when we remove the code API as well.

What did we do with other cases?

Reply:
If it's the first time it's deprecated, yes. I would even be more clear and add:

```
IMPORTANT: This feature will be removed soon, please update your code base
```

Reply:
Sure.